Fix RVA field access in composite Crossgen2 build (#78723)

trylek opened the pull request:
The pre-existing behavior was copying all RVA fields in single-file builds and only selectively copying explicitly referenced fields in composite builds. This has bitten us in WPF where Managed C++ emits an array of early initializers that are marked by an initial and final RVA field. By only selectively copying those two without the remaining entries in-between we end up crashing later on as the app is not fully initialized.

Based on David Wrighton's advice I have kept the logic for copying RVA fields in case of array initializers that seem to be hot in terms of performance. For the remaining field access I have ended up constructing a new precode fixup RVA_FIELD_ADDRESS for this purpose.

Fixes: #78723

Thanks

Tomas

/cc @dotnet/crossgen-contrib

trylek committed:
Fix RVA field access in composite Crossgen2 build (#78723)

The pre-existing behavior was copying all RVA fields in
single-file builds and only selectively copying explicitly
referenced fields in composite builds. This has bitten us
in WPF where Managed C++ emits an array of early initializers
that are marked by an initial and final RVA field. By only
selectively copying those two without the remaining entries
in-between we end up crashing later on as the app is not
fully initialized.

Based on David Wrighton's advice I have kept the logic for
copying RVA fields in case of array initializers that seem
to be hot in terms of performance. For the remaining field
access I have ended up constructing a new precode fixup
RVA_FIELD_ADDRESS for this purpose.

Fixes: #78723

Thanks

Tomas

jkotas commented on src/coreclr/tools/Common/Internal/Runtime/ReadyToRunConstants.cs:
```diff
@@ -165,6 +165,8 @@ public enum ReadyToRunFixupKind
         Check_IL_Body              = 0x35, /* Check to see if an IL method is defined the same at runtime as at compile time. A failed match will cause code not to be used. */
         Verify_IL_Body             = 0x36, /* Verify an IL body is defined the same at compile time and runtime. A failed match will cause a hard runtime failure. */
 
+        RvaFieldAddress            = 0x37, /* Used in Crossgen2 to address RVA fields in their original MSIL images */
```
Can we just use existing `FieldAddress` fixup for this?

trylek replied:
I hoped we could but it looks like <code>FieldAddress</code> is specific to instance fields and cannot be used as a precode. The newly introduced fixup is the most similar to <code>StaticFieldAddress</code> but that one does something weird around storing the result next to the fixup cell, I guess that's something specific to runtime JIT behavior.

trylek replied:
Please see

https://github.com/dotnet/runtime/blob/f924d6b5f69792f33a507b47a1fc870a185ad638/src/coreclr/vm/jitinterface.cpp#L13825

jkotas replied:
I do not see the problem. The comment on `FieldAddress` says that it is for static fields:

FieldAddress = 0x20, // For accessing a cross-module static fields

If FieldAddress cannot be used as precode fixup, we can make it usable as precode fixup. We have other fixups like that. The encoding is designed to allow the same `ReadyToRunFixupKind` to be used in different modes.

trylek replied:
OK, thanks for the feedback, I'll take look into making <code>FieldAddress</code> work for this purpose.

jkotas replied:
> Please see
> https://github.com/dotnet/runtime/blob/f924d6b5f69792f33a507b47a1fc870a185ad638/src/coreclr/vm/jitinterface.cpp#L13825

You can delete `ENCODE_STATIC_FIELD_ADDRESS` while you are on it. It looks like a left-over from fragile NGen.

trylek replied:
Fixed in 2nd commit, thanks for the suggestions!

trylek committed:
Address JanK's PR feedback

trylek commented:
/azp run runtime-coreclr outerloop

jkotas commented on src/coreclr/inc/corcompile.h:
```diff
@@ -161,10 +161,10 @@ enum CORCOMPILE_FIXUP_BLOB_KIND
     ENCODE_MODULE_HANDLE                = 0x50,     /* Module token */
-    ENCODE_STATIC_FIELD_ADDRESS,                    /* For accessing a static field */
-    ENCODE_MODULE_ID_FOR_GENERIC_STATICS,           /* For accessing static fields */
+    // obsolete: ENCODE_STATIC_FIELD_ADDRESS = 0x51, /* For accessing a static field */
+    ENCODE_MODULE_ID_FOR_GENERIC_STATICS = 0x52,    /* For accessing static fields */
```
```suggestion
    ENCODE_MODULE_ID_FOR_GENERIC_STATICS, /* For accessing static fields */
```
These are not part of the stable R2R format. They should be all deleted (it can be separate change).

jkotas committed:
Update src/coreclr/inc/corcompile.h

jkotas commented on src/coreclr/inc/corcompile.h:
```diff
@@ -161,9 +161,8 @@ enum CORCOMPILE_FIXUP_BLOB_KIND
     ENCODE_CHECK_IL_BODY,                           /* Check to see if an IL method is defined the same at runtime as at compile time. A failed match will cause code not to be used. */
     ENCODE_VERIFY_IL_BODY,                          /* Verify an IL body is defined the same at compile time and runtime. A failed match will cause a hard runtime failure. */
-
+    
```
```suggestion

```

jkotas committed:
Update src/coreclr/inc/corcompile.h

jkotas commented on src/coreclr/inc/corcompile.h:
```diff
@@ -161,9 +161,7 @@ enum CORCOMPILE_FIXUP_BLOB_KIND
 
     ENCODE_CHECK_IL_BODY,                           /* Check to see if an IL method is defined the same at runtime as at compile time. A failed match will cause code not to be used. */
     ENCODE_VERIFY_IL_BODY,                          /* Verify an IL body is defined the same at compile time and runtime. A failed match will cause a hard runtime failure. */
```
```suggestion
    ENCODE_VERIFY_IL_BODY,                          /* Verify an IL body is defined the same at compile time and runtime. A failed match will cause a hard runtime failure. */

```

jkotas committed:
Update src/coreclr/inc/corcompile.h

jkotas reviewed (approved):
LGTM!

trylek commented:
/azp run runtime-coreclr outerloop

trylek committed:
Revert unintentional change of one other place involving statics

I have noticed that, probably as part of bulk renames, I have
erroneously changed one other place in getFieldInfo to use the
new helper RvaFieldAddress; this is not appropriate as the place
doesn't deal with RVA fields. This was the reason why several
readytorun/tests failed in the PR testing.

Thanks

Tomas

trylek commented:
/azp run runtime-coreclr outerloop

trylek committed:
Update TestRVAField that no longer fails to JIT compile

With the RVA fix we can now precompile the TestRVAField test case
just fine.

Thanks

Tomas

trylek commented:
/azp run runtime-coreclr outerloop

trylek commented:
The outerloop failure is known, https://github.com/dotnet/runtime/issues/85663, merging in. I'll be happy to address any additional feedback in a follow-up change. I'll also put up a PR for blocking the test in issues.targets to make outerloop clean again.
